transports/tls-quic/src/certificate.rs

```diff
@@ -0,0 +1,231 @@
+/// certificate extension containing the public key of the given keypair.
+pub fn make_certificate(
+    keypair: &identity::Keypair,
+) -> Result<rcgen::Certificate, rcgen::RcgenError> {
+
+    // Certificates MUST use the NamedCurve encoding for elliptic curve parameters.
+    // Similarly, hash functions with an output length less than 256 bits MUST NOT be used
+    let signature_algorithm = &rcgen::PKCS_ECDSA_P256_SHA256;
+
+    // The key used to generate and sign this certificate
+    // SHOULD NOT be related to the host's key.
+    // Endpoints MAY generate a new key and certificate
+    // for every connection attempt, or they MAY reuse the same key
+    // and certificate for multiple connections.
+    let certif_keypair = rcgen::KeyPair::generate(signature_algorithm)?;
+
+    // The certificate MUST contain the libp2p Public Key Extension.
+    let libp2p_extension: rcgen::CustomExtension = {
+
+        // The peer signs the concatenation of the string `libp2p-tls-handshake:`
+        // and the public key that it used to generate the certificate carrying
+        // the libp2p Public Key Extension, using its private host key.
+        let signature = {
+            let mut msg = vec![];
+            msg.extend(*b"libp2p-tls-handshake:");
+            msg.extend(certif_keypair.public_key_der());
+
+            keypair.sign(&msg).expect("Let's hope the signing won't fail")
+        };
+
+        // The public host key and the signature are ANS.1-encoded
+        // into the SignedKey data structure, which is carried
+        // in the libp2p Public Key Extension.
+        // SignedKey ::= SEQUENCE {
+        //    publicKey BIT STRING,
+        //    signature BIT STRING
+        // }
+        let extension_content = {
+            let serialized_pubkey = keypair.public().into_protobuf_encoding();
+            yasna::encode_der(&(serialized_pubkey, signature))
+        };
+
+        // The libp2p Public Key Extension is a X.509 extension
+        // with the Object Identier 1.3.6.1.4.1.53594.1.1,
+        // allocated by IANA to the libp2p project at Protocol Labs.
+        let libp2p_ext_oid = &[1, 3, 6, 1, 4, 1, 53594, 1, 1];  // Based on libp2p TLS 1.3 specs
+
+        // This extension MAY be marked critical.
+        rcgen::CustomExtension::from_oid_content(libp2p_ext_oid, extension_content)
+    };
+
+    let certificate = {
+        let mut params = rcgen::CertificateParams::new(vec![]);
+        params.distinguished_name = rcgen::DistinguishedName::new();
+        params.custom_extensions.push(libp2p_extension);
+        params.alg = signature_algorithm;
+        params.key_pair = Some(certif_keypair);
+        rcgen::Certificate::from_params(params)?
+    };
+
+    Ok(certificate)
+}
+
+// The contents of the specific libp2p extension, containing the public host key
+// and a signature performed using the private host key
+pub struct P2pExtension {
+    public_key: identity::PublicKey,
+    signature: Vec<u8>,
```

Comment:
I suggest wrapping these in a newtypes to prevent from accidental swapping.

Reply:
Nothing bad should happen if we swap them. They are of different types and there is no ser/de involved.